xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,402 @@
+		gsb.currentMu.Lock()
+		defer gsb.currentMu.Unlock()
+		cur.Close()
+	}()
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) balancerCurrentOrPending(bw *balancerWrapper) bool {
+	return bw == gsb.balancerCurrent || bw == gsb.balancerPending
+}
+
+// caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) newSubConn(bw *balancerWrapper, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if !gsb.balancerCurrentOrPending(bw) {
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bw.Balancer, bw.Balancer)
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	return sc, nil
+}
+
+func (gsb *gracefulSwitchBalancer) resolveNow(bw *balancerWrapper, opts resolver.ResolveNowOptions) {
+	// If the resolver specifies an update with a whole new type of policy,
+	// there is no reason to forward a ResolveNow call from the balancer being
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	if bw == gsb.balancerPending || (bw == gsb.balancerCurrent && gsb.balancerPending == nil) {
+		gsb.cc.ResolveNow(opts)
+	}
+}
+
+// caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) removeSubConn(bw *balancerWrapper, sc balancer.SubConn) {
+	if !gsb.balancerCurrentOrPending(bw) {
+		return
+	}
+	gsb.cc.RemoveSubConn(sc)
+}
+
+func (gsb *gracefulSwitchBalancer) updateAddresses(bw *balancerWrapper, sc balancer.SubConn, addrs []resolver.Address) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bw) {
+		return
+	}
+	gsb.cc.UpdateAddresses(sc, addrs)
+}
+
+// SwitchTo gracefully switches to the new balancer. This function must be
+// called synchronously alongside the rest of the balancer.Balancer methods this
+// Graceful Switch Balancer implements.
+func (gsb *gracefulSwitchBalancer) SwitchTo(builder balancer.Builder) error {
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		return errBalancerClosed
+	}
+	bw := &balancerWrapper{
+		gsb:       gsb,
+		lastState: defaultState,
+		subconns:  make(map[balancer.SubConn]bool),
+	}
+	var balToClose balancer.Balancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = bw
+	} else {
+		// Clean up resources here that are from a previous pending lb.
+		if gsb.balancerPending != nil {
+			for sc := range gsb.balancerPending.subconns {
+				gsb.cc.RemoveSubConn(sc)
+			}
+		}
+		if gsb.balancerPending != nil {
+			balToClose = gsb.balancerPending
+		}
+		gsb.balancerPending = bw
+	}
+	gsb.mu.Unlock()
+
+	newBalancer := builder.Build(bw, gsb.bOpts)
+	if newBalancer == nil {
+		if balToClose != nil {
+			balToClose.Close()
+		}
+		if gsb.balancerPending != nil {
+			gsb.balancerPending = nil
+		} else {
+			gsb.balancerCurrent = nil
+		}
+		return balancer.ErrBadResolverState
+	}
+	bw.Balancer = newBalancer
```

Comment:
Add a comment why this doesn't need to take the mu.

Reply:
Added - "This write doesn't need to take gsb.mu because this field never gets read or written to on any calls from the current or pending. Calls from grpc to this balancer are guaranteed to be called synchronously, so this bw.Balancer field will never be forwarded to until this SwitchTo() function returns."